### PR TITLE
Feature motion planning manager

### PR DESCRIPTION
**Pull request by jrgnicho**
This PR implements parts of the motion planning manager to allows calling the process and free-space motion planners as needed.  
- Free-space motion planner node:
  - Can take a joint goal as well.
  - Parameterized cartesian weight coefficients in case it's necessary to tweak those.
  - A few other minor cosmetic changes

NOTE: Previewing the motion plans is somewhat involved and so it will be done in a subsequent PR

### CODE REVIEW

**Comment by jrgnicho** [comment]
@marrts I've rebased the master, it's ready for review

File: crs_application/src/task_managers/motion_planning_manager.cpp [hunk 1]
```diff
@@ -33,8 +33,23 @@
...
 
+#include <tf2_eigen/tf2_eigen.h>
+#include <boost/optional.hpp>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include <trajectory_msgs/msg/joint_trajectory.hpp>
 #include "crs_application/task_managers/motion_planning_manager.h"
 
+static const double WAIT_SERVICE_DURATION = 2.0; // secs
+static const double WAIT_SERVICE_COMPLETION_PERIOD = 30.0; // secs
+static const std::string CALL_FREESPACE_MOTION_SERVICE = "plan_free_space";
```
Comment (marrts) [inline crs_application/src/task_managers/motion_planning_manager.cpp]: This freespace motion service name should match the one in the freespace planning service node

File: crs_application/src/task_managers/motion_planning_manager.cpp [hunk 2]
```diff
@@ -53,37 +68,246 @@ MotionPlanningManager::~MotionPlanningManager()
...
+
+      double d = (p2 - p1).norm();
+      if(current_dist + d > split_dist)
+      {
+        RCLCPP_INFO(node_->get_logger(),"Found split at point %lu of raster %lu",p_idx, r_idx);
+        split_point_idx = p_idx;
+        current_dist = 0;
+        break;
+      }
+      current_dist += d;
```
Comment (marrts) [inline crs_application/src/task_managers/motion_planning_manager.cpp]: I think the current_dist+=d should be in an else statement. This distance won't be added if the next point is the first after a split. Although this difference will likely be relatively minor overall.

File: crs_application/src/task_managers/motion_planning_manager.cpp [hunk 3]
```diff
@@ -53,37 +68,246 @@ MotionPlanningManager::~MotionPlanningManager()
...
+  // splitting paths
+  std::vector<datatypes::ProcessToolpathData> toolpaths_processes;
+  double current_dist = 0.0;
+  Eigen::Vector3d p1, p2;
+  toolpaths_processes.resize(1);
+  for(std::size_t r_idx = 0 ; r_idx < input_->rasters.size(); r_idx++)
+  {
+    geometry_msgs::msg::PoseArray& raster = input_->rasters[r_idx];
+    int split_point_idx = -1;
+    for(std::size_t p_idx = 0; p_idx < raster.poses.size() - 1; p_idx++)
```
Comment (marrts) [inline crs_application/src/task_managers/motion_planning_manager.cpp:171]: I think this fails to account for the remaining distance in a raster if a split occurs and instead will pickup at the beginning of the next raster with a distance of zero.

File: crs_application/src/task_managers/motion_planning_manager.cpp [hunk 4]
```diff
@@ -53,37 +68,246 @@ MotionPlanningManager::~MotionPlanningManager()
...
+  {
+    msg::ToolProcessPath process_path;
+    for(const geometry_msgs::msg::PoseArray& raster : pd.rasters)
+    {
+      process_path.rasters.push_back(raster);
+    }
+    req->process_paths.push_back(process_path);
+  }
+
+  std::shared_future<srv::PlanProcessMotions::Response::SharedPtr> fut = process_motion_planning_client_->async_send_request(
```
Comment (marrts) [inline crs_application/src/task_managers/motion_planning_manager.cpp:271]: The request message is only being populated with process paths, the config settings approach dist, retreat dist, etc should also be added to the request message

**Review by marrts — CHANGES_REQUESTED** [review]
Mostly looks good, just a few small things I noticed in the motion planning manager

File: crs_application/src/task_managers/motion_planning_manager.cpp (hunk 3 above)
Comment (jrgnicho) [reply crs_application/src/task_managers/motion_planning_manager.cpp:171]: I think it captures the remaining of the raster inside the if statement
```
    if(split_point_idx >=0)
    {
    .
    .
    .
```

**Comment by jrgnicho** [comment]
@marrts your comments/change requests have been addressed

File: crs_application/src/task_managers/motion_planning_manager.cpp [hunk 5]
```diff
@@ -149,33 +177,52 @@ common::ActionResult MotionPlanningManager::splitToolpaths()
...
+    for(std::size_t p_idx = 1; p_idx < breakpoints_indices.size(); p_idx++)
+    {
+      // copy portion of raster leading up to breakpoint into current toolpath
+      end_idx = breakpoints_indices[p_idx];
       new_raster.poses.clear();
-      std::copy(raster.poses.begin() + split_point_idx + 1, raster.poses.end(),
+      std::copy(raster.poses.begin() + start_idx, raster.poses.begin() + end_idx,
                 std::back_inserter(new_raster.poses));
-      toolpaths_processes.back().rasters.push_back(new_raster);
+      toolpaths_processes.back().rasters.push_back(raster);
```
Comment (marrts) [inline crs_application/src/task_managers/motion_planning_manager.cpp]: I believe you meant to put 
```suggestion
      toolpaths_processes.back().rasters.push_back(new_raster);
```

File: crs_application/src/task_managers/motion_planning_manager.cpp [hunk 6]
```diff
@@ -149,33 +177,52 @@ common::ActionResult MotionPlanningManager::splitToolpaths()
...
-    else
+
+    // copy remaining segment of raster into toolpath
+    if(end_idx < raster.poses.size()-1)
     {
-      // no split so copy current raster into current process
+      new_raster.poses.clear();
+      std::copy(raster.poses.begin() + end_idx, raster.poses.end(),
+                std::back_inserter(new_raster.poses));
       toolpaths_processes.back().rasters.push_back(raster);
```
Comment (marrts) [inline crs_application/src/task_managers/motion_planning_manager.cpp]: This line should also be changed to new_raster

File: crs_application/src/task_managers/motion_planning_manager.cpp [hunk 7]
```diff
@@ -149,33 +177,52 @@ common::ActionResult MotionPlanningManager::splitToolpaths()
...
       new_raster.poses.clear();
-      std::copy(raster.poses.begin() + split_point_idx + 1, raster.poses.end(),
+      std::copy(raster.poses.begin() + start_idx, raster.poses.begin() + end_idx,
                 std::back_inserter(new_raster.poses));
-      toolpaths_processes.back().rasters.push_back(new_raster);
+      toolpaths_processes.back().rasters.push_back(raster);
+
+      // create new toolpath
+      toolpaths_processes.resize(toolpaths_processes.size() + 1);
+      start_idx = p_idx;
```
Comment (marrts) [inline crs_application/src/task_managers/motion_planning_manager.cpp]: I think this should be
```suggestion
      start_idx = breakpoints_indices[p_idx];
```

**Review by marrts — CHANGES_REQUESTED** [review]
Looks good, I think you just had a couple typos referencing the wrong variable. Also make sure #16 changes are merged in so that this builds with the latest tesseract changes.

File: crs_application/src/task_managers/motion_planning_manager.cpp (hunk 5 above)
Comment (jrgnicho) [reply crs_application/src/task_managers/motion_planning_manager.cpp]: yes good catch

**Comment by jrgnicho** [comment]
@marrts made the requested changes and rebased, please review, thanks

**Review by marrts — APPROVED** [review]
Looks good to me

**Pull request merged**